File: ppc/wiki-game/solve.py
T = int(input())

for _ in range(T):
    n, m = map(int, input().split())
    adj = [[] for _ in range(n)]
    for i in range(m):
        u, v = map(int, input().split())
        adj[u].append(v)

    src, dst = map(int, input().split())
    visited = set({src})
    for i in range(6):
        cur = list(visited)
        for c in cur:
            for nxt in adj[c]:
                visited.add(nxt)

    if dst in visited:
        print("YES")
    else:
        print("NO")

File: ppc/wiki-game/solve.js
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

let T = -1;
let n = -1, m = -1, cnt;
let edges;

const bfs = (src, dst) => {
    let queue = [[src, 0]];
    let visited = new Array(n).fill(false);

    while (queue.length !== 0) {
        let [u, d] = queue.shift();

        if (u === dst && d <= 6) {
            return 'YES';
        }

        visited[u] = true;

        for (let v of edges[u]) {
            if (!visited[v]) {
                queue.push([v, d + 1]);
            }
        }
    }

    return 'NO';
}

rl.on('line', line => {
    if (T === -1) {
        T = parseInt(line);
    } else if (n === -1) {
        [n, m] = line.trim().split(' ').map(x => parseInt(x));
        edges = new Array(n).fill(null).map(_ => []);
        cnt = 0;
    } else if (cnt === m) {
        let [src, dst] = line.trim().split(' ').map(x => parseInt(x));
        console.log(bfs(src, dst));
        n = -1;
        T -= 1;

        if (T === 0) {
            rl.close();
        }
    } else {
        let [u, v] = line.trim().split(' ').map(x => parseInt(x));
        edges[u].push(v);
        cnt += 1;
    }
});